Initial (#1)
RangerMauve/ipld-prolly-indexer: README.md

Export the proof of a record and related blocks

### `func (collection *Collection) Get(ctx context.Context, recordId []byte) (ipld.Node, error)`
Get data by the recordId

### `func (collection *Collection) Indexes(ctx context.Context) ([]Index, error)`
Return all indexes of the collection

### `func (collection *Collection) Insert(ctx context.Context, record ipld.Node) error`
Insert a record into DB including raw data and index creating

### `func (db *Database) GetDBMetaInfo() (*schema.DBMetaInfo, error)`
Get metainfo about the database

### `func (record *Record) AsIPLDWithProof(proof tree.Proof) (ipld.Node, error)`
Encode record and its proof as ipld node.

### `func (db *Database) RootCid() cid.Cid`

### `func (db *Database) GetBlockstore() *blockstore.Blockstore`

### `func (index *Index) Fields() []string`

### `func (index *Index) Exists() bool`

### `func (collection *Collection) BestIndex(ctx context.Context, query Query) (*Index, error)`
Choose best index for a query

### `func (query Query) Matches(record Record) bool`
Return whether the record matches the query or not

### `func (record *Record) AsIPLD() (datamodel.Node, error)`
Return record as IPLD node

> ### `func ParseStringsFromCBOR(data []byte) ([]string, error)`

Author: These last ones should be removed from the docs since they're only relevant to advanced uses

This field will be increased for new index types and versions.
Additional properties may be added in the future.

The fields being indexed are represented in a dag-cbor array and this is used as the "name" of the index.

#### `\x00{collection utf8 name}\x00i\x00{dag-cbor array of field names}\x00{index key bytes}`

Indexed record keys. The fields from the record will be pulled out and turned into a dag-cbor array containing the fields, and the record ID.

This will point to a value of the primary key for the record.

The primary key can then be used to resolve the record.

## Data Struct

```Golang
type Database struct {
	blockStore blockstore.Blockstore
	linkSystem  *ipld.LinkSystem
	rootCid     cid.Cid
	tree        *tree.ProllyTree
	collections map[string]*Collection
}
```
 linkSystem and prollyTree are based on the blockstore, root cid is included in tree.

```Golang
type Collection struct {
	db         *Database
	name       string
	primaryKey []string
}
```

```Golang
type Index struct {
	collection *Collection
	fields     []string
}
```

```Golang
type Query struct {
	Equal   map[string]ipld.Node
	Compare *CompareCondition
	Sort    string
	Limit   int
	Skip    int
}
```

> If here are some indexed fields in the equal map, search will choose the best index to find the recordId.If the equal map is nil, search will iterate all records and find records matche the query.

Author: ```suggestion

If there are indexed fields in the equal map, search will choose the best index where the field is at the start of the fields. If the equal map is nil, search will iterate all records and filter out records that don't match the query.
```

```Golang
type InclusionProof struct {
	Key   []byte
	Proof tree.Proof
	Root  cid.Cid
}
```

> Proof is the ProllyTree path of the record, root cid is cid of ProllyTree's root.It should be noted that you may get different InclusionProofs for a record in different time because the ProllyTree may be updated.

Author: ```suggestion

Proof is the ProllyTree path of the record, root cid is cid of ProllyTree's root.It should be noted that you may get different InclusionProofs for a record in different time because the ProllyTree may be updated.
```
